src/http2_frame.erl

```diff
@@ -217,20 +233,43 @@ header_to_binary(#frame_header{
     <<L:24,T:8,F:8,0:1,StreamId:31>>.
 
 -spec payload_to_binary(payload()) -> {frame_type(), iodata()}.
-payload_to_binary(P=#data{})          -> {?DATA, http2_frame_data:to_binary(P)};
-payload_to_binary(P=#headers{})       -> {?HEADERS, http2_frame_headers:to_binary(P)};
-payload_to_binary(P=#priority{})      -> {?PRIORITY, http2_frame_priority:to_binary(P)};
-payload_to_binary(P=#rst_stream{})    -> {?RST_STREAM, http2_frame_rst_stream:to_binary(P)};
-payload_to_binary(P=#settings{})      -> {?SETTINGS, http2_frame_settings:to_binary(P)};
-payload_to_binary(P=#push_promise{})  -> {?PUSH_PROMISE, http2_frame_push_promise:to_binary(P)};
-payload_to_binary(P=#ping{})          -> {?PING, http2_frame_ping:to_binary(P)};
-payload_to_binary(P=#goaway{})        -> {?GOAWAY, http2_frame_goaway:to_binary(P)};
-payload_to_binary(P=#window_update{}) -> {?WINDOW_UPDATE, http2_frame_window_update:to_binary(P)};
-payload_to_binary(P=#continuation{})  -> {?CONTINUATION, http2_frame_continuation:to_binary(P)}.
+payload_to_binary(P) ->
+    Type = payload_type(P),
+
+    Bin =
+        case Type of
+            ?DATA -> http2_frame_data:to_binary(P);
+            ?HEADERS -> http2_frame_headers:to_binary(P);
+            ?PRIORITY -> http2_frame_priority:to_binary(P);
+            ?RST_STREAM -> http2_frame_rst_stream:to_binary(P);
+            ?SETTINGS ->  http2_frame_settings:to_binary(P);
+            ?PUSH_PROMISE -> http2_frame_push_promise:to_binary(P);
+            ?PING -> http2_frame_ping:to_binary(P);
+            ?GOAWAY -> http2_frame_goaway:to_binary(P);
+            ?WINDOW_UPDATE -> http2_frame_window_update:to_binary(P);
+            ?CONTINUATION -> http2_frame_continuation:to_binary(P)
+        end,
+    {Type, Bin}.
 
 iodata_size(L) when is_list(L) ->
     lists:foldl(fun(X, Acc) ->
                         Acc + iodata_size(X)
                 end, 0, L);
 iodata_size(B) when is_binary(B) ->
     byte_size(B).
+
+%% Breaking my own abstraction here, but only for a performance
+%% optimization. This function assumes that records are built with the
+%% record name as the first element in the tuple, and if that changes,
+%% or if payloads are no longer records, then this will stop working.
+-spec payload_type(payload()) -> frame_type().
+payload_type(P) when element(1, P) =:= data          -> ?DATA;
+payload_type(P) when element(1, P) =:= headers       -> ?HEADERS;
+payload_type(P) when element(1, P) =:= priority      -> ?PRIORITY;
+payload_type(P) when element(1, P) =:= rst_stream    -> ?RST_STREAM;
+payload_type(P) when element(1, P) =:= settings      -> ?SETTINGS;
+payload_type(P) when element(1, P) =:= push_promise  -> ?PUSH_PROMISE;
+payload_type(P) when element(1, P) =:= ping          -> ?PING;
+payload_type(P) when element(1, P) =:= goaway        -> ?GOAWAY;
+payload_type(P) when element(1, P) =:= window_update -> ?WINDOW_UPDATE;
+payload_type(P) when element(1, P) =:= continuation  -> ?CONTINUATION.
```

Comment:
why even have the constants, actually? you could just use the name of each module, then you could just call everything directly with `Type:to_binary` or whatever